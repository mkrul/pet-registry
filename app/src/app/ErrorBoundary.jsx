import { Component } from "react";

class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false, error: null };
  }

  static getDerivedStateFromError(error) {
    return { hasError: true, error };
  }

  componentDidCatch(error, errorInfo) {
    console.error("ðŸš¨ ErrorBoundary: Error caught:", error, errorInfo);
    console.error("ðŸš¨ ErrorBoundary: Error stack:", error.stack);
  }

  handleClose = () => {
    this.setState({ hasError: false, error: null });
  };

  render() {
    if (this.state.error) {
      return (
        <div
          className="bg-red-100 border-red-400 text-red-700 border px-4 py-3 rounded relative mb-4 flex justify-between items-center"
          role="alert"
        >
          <span className="block sm:inline">{this.state.error.message}</span>
          <button onClick={this.handleClose} className="ml-4" aria-label="Close">
            <span aria-hidden="true" className="text-xl">
              Ã—
            </span>
          </button>
        </div>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
