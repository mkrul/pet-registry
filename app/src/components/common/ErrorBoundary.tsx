import React, { Component, ErrorInfo, ReactNode } from "react";
import Notification from "../shared/Notification";
import { NotificationType } from "../../types/shared/Notification";

interface Props {
  children: ReactNode;
}

interface State {
  hasError: boolean;
  error: Error | null;
}

class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { hasError: false, error: null };
  }

  static getDerivedStateFromError(error: Error): State {
    return { hasError: true, error };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    console.error("ErrorBoundary caught an error:", error, errorInfo);
  }

  render() {
    if (this.state.hasError && this.state.error) {
      return (
        <Notification
          type={NotificationType.ERROR}
          message={this.state.error.message}
          onClose={() => this.setState({ hasError: false, error: null })}
        />
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
