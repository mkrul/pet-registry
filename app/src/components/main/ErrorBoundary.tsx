import { Component, ErrorInfo } from "react";
import Notification from "../common/Notification";
import { NotificationType } from "../../types/common/Notification";
import { ErrorBoundaryProps, ErrorBoundaryState } from "../../types/main/ErrorBoundary";

class ErrorBoundary extends Component<ErrorBoundaryProps, ErrorBoundaryState> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false, error: null };
  }

  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    return { hasError: true, error };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    console.error("ErrorBoundary caught an error:", error, errorInfo);
  }

  handleClose = () => {
    this.setState({ hasError: false, error: null });
  };

  render() {
    if (this.state.hasError && this.state.error) {
      return (
        <Notification
          type={NotificationType.ERROR}
          message={this.state.error.message}
          onClose={this.handleClose}
        />
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
