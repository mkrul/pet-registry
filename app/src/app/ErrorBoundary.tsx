import { Component, ErrorInfo, ReactNode } from "react";
import { ErrorBoundaryProps, ErrorBoundaryState } from "../shared/types/main/ErrorBoundary";

class ErrorBoundary extends Component<ErrorBoundaryProps, ErrorBoundaryState> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false, error: null };
  }

  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    return { hasError: true, error };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo): void {
    console.error("Error caught by ErrorBoundary:", error, errorInfo);
  }

  handleClose = () => {
    this.setState({ hasError: false, error: null });
  };

  render(): ReactNode {
    if (this.state.error) {
      return (
        <div
          className="bg-red-100 border-red-400 text-red-700 border px-4 py-3 rounded relative mb-4 flex justify-between items-center"
          role="alert"
        >
          <span className="block sm:inline">{this.state.error.message}</span>
          <button onClick={this.handleClose} className="ml-4" aria-label="Close">
            <span aria-hidden="true" className="text-xl">
              Ã—
            </span>
          </button>
        </div>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
