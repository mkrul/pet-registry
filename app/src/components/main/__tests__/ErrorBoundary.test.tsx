import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import ErrorBoundary from "../ErrorBoundary";

// Mock component that throws an error
const ThrowError = ({ shouldThrow = false }) => {
  if (shouldThrow) {
    throw new Error("Test error");
  }
  return <div>Normal Component</div>;
};

// Mock console.error to prevent error logging during tests
const originalError = console.error;
beforeEach(() => {
  console.error = vi.fn();
});

afterEach(() => {
  console.error = originalError;
  cleanup();
});

describe("ErrorBoundary", () => {
  describe("Normal Rendering", () => {
    it("renders children when no error occurs", () => {
      const { container } = render(
        <ErrorBoundary>
          <ThrowError shouldThrow={false} />
        </ErrorBoundary>
      );

      expect(screen.getByText("Normal Component")).toBeDefined();
      expect(container).toMatchSnapshot();
    });
  });

  describe("Error Handling", () => {
    it("displays error notification when error occurs", () => {
      const { container } = render(
        <ErrorBoundary>
          <ThrowError shouldThrow={true} />
        </ErrorBoundary>
      );

      const errorAlert = screen.getByRole("alert");
      expect(errorAlert).toBeDefined();
      expect(screen.getByText("Test error")).toBeDefined();
      expect(container).toMatchSnapshot();
    });

    it("can dismiss error notification", () => {
      const { unmount, rerender } = render(
        <ErrorBoundary>
          <ThrowError shouldThrow={true} />
        </ErrorBoundary>
      );

      const closeButton = screen.getByRole("button", { name: /close/i });
      expect(closeButton).toBeDefined();

      fireEvent.click(closeButton);
      unmount();

      // Re-render with new content after error is cleared
      render(
        <ErrorBoundary>
          <div>Recovered Content</div>
        </ErrorBoundary>
      );

      expect(screen.queryByRole("alert")).toBeNull();
      expect(screen.getByText("Recovered Content")).toBeDefined();
    });
  });

  describe("Error Message Display", () => {
    it("shows error message consistently across environments", () => {
      const environments = ["development", "production"];
      const originalEnv = process.env.NODE_ENV;

      environments.forEach(env => {
        cleanup();
        process.env.NODE_ENV = env;
        render(
          <ErrorBoundary>
            <ThrowError shouldThrow={true} />
          </ErrorBoundary>
        );

        const errorElements = screen.getAllByText("Test error");
        expect(errorElements.length).toBe(1);
        expect(screen.getByRole("alert")).toBeDefined();
        cleanup();
      });

      process.env.NODE_ENV = originalEnv;
    });
  });
});
