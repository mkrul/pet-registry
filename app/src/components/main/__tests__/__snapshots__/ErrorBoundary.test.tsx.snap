// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorBoundary > Error Handling > displays error notification when error occurs 1`] = `
<div>
  <div
    class="bg-red-100 border-red-400 text-red-700 border px-4 py-3 rounded relative mb-4 flex justify-between items-center"
    role="alert"
  >
    <span
      class="block sm:inline"
    >
      Test error
    </span>
    <button
      aria-label="Close"
      class="ml-4"
    >
      <span
        aria-hidden="true"
        class="text-xl"
      >
        Ã—
      </span>
    </button>
  </div>
</div>
`;

exports[`ErrorBoundary > Normal Rendering > renders children when no error occurs 1`] = `
<div>
  <div>
    Normal Component
  </div>
</div>
`;
